lib/rpc/status.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rpc
+
+import (
+	"context"
+
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/defaults"
+	"github.com/gravitational/gravity/lib/modules"
+	"github.com/gravitational/gravity/lib/storage"
+	"github.com/gravitational/gravity/lib/utils"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// GravityAgentStatus contains a gravity agent's status information.
+type GravityAgentStatus struct {
+	// Node specifies the name of the node running the agent.
+	Node string
+	// IP specifies the IP address of the node running the agent.
+	IP string
+	// Status indiciates the current status of the agent. An agent is `Deployed`
+	// if the gravity-agent service is active. The agent is `Offline` if it
+	// fails to respond to the status request.
+	Status string
+	// Version describes gravity agent version.
+	Version string
+}
+
+func AgentStatus(ctx context.Context, servers []string, logger log.FieldLogger, rpc AgentRepository) (
+	[]GravityAgentStatus, error) {
+	errs := make(chan error, len(servers))
+	statusCh := make(chan GravityAgentStatus, len(servers))
+	for _, srv := range servers {
+		go func(host string) {
+			systemInfo, err := getSystemInfo(ctx, host, rpc)
```

Comment:
I would encapsulate these two (getSystemInfo and getVersion) into a separate function (getAgentStatus) just so the error handling is a little easier, for example:

```go
func getAgentStatus(ctx context.Context, address string, rpc AgentRepository) AgentStatus {
  systemInfo, err := getSystemInfo(ctx, host, rpc)
  // return status with Error set
  version, err := getVersion(ctx, host, rpc)
  // return status with Error set
  return AgentStatus{
    Address:  address,
    Status:   constants.GravityAgentDeployed,
    Hostname: systemInfo.GetHostname(),
    Version:  version.Version,    
  }, nil
}

// Then the goroutine becomes trivial:

go func(address string) {
  statusCh <- getAgentStatus(ctx, address, rpc)
}
```

Reply:
Realized we don't even need to request systemInfo since we are only collecting hostname from it. We can pull hostname from the cluster state.